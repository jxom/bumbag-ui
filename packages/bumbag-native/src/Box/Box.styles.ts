import { styled } from '../styled';
import { theme } from '../utils/theme';

export const StyledBox = styled.View`
  ${theme('native.Box', 'styles.base')};
`;

export const StyledBoxSafe = styled.SafeAreaView`
  ${theme('native.Box.Safe', 'styles.base')};
`;

export const StyledBoxScroll = styled.ScrollView`
  ${theme('native.Box.Scroll', 'styles.base')};
`;

export const StyledBoxKeyboardAvoiding = styled.KeyboardAvoidingView`
  ${theme('native.Box.KeyboardAvoiding', 'styles.base')};
`;

export const StyledBoxTouchable = styled.TouchableOpacity`
  ${theme('native.Box.Touchable', 'styles.base')};
`;

export const StyledBoxTouchableWithoutFeedback = styled.TouchableWithoutFeedback`
  ${theme('native.Box.TouchableWithoutFeedback', 'styles.base')};
`;
