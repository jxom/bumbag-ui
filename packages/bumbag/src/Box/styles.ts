import { css, cssClass } from '../styled';
import { altitude, theme } from '../utils/theme';

const FLEX_HORIZONTAL_ALIGN_MAP = {
  left: 'flex-start',
  center: 'center',
  right: 'flex-end',
};

const FLEX_VERTICAL_ALIGN_MAP = {
  top: 'flex-start',
  center: 'center',
  bottom: 'flex-end',
};

export const style = (styleProps) => cssClass`
  & {
    ${styleProps.style};
  }

  ${getAlignmentAttributes(styleProps)};

  ${
    styleProps.altitude &&
    css`
      & {
        ${altitude(styleProps.altitude)(styleProps)};
      }
    `
  };
`;

export const Box = (styleProps) => cssClass`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  & {
    ${theme(styleProps.themeKey, 'css.root')(styleProps)};
  }
`;

export function getAlignmentAttributes(styleProps) {
  return css`
    ${(styleProps.alignY || styleProps.alignX) &&
    css`
      display: flex;

      ${
        !styleProps.display &&
        css`
          flex-direction: column;
        `
      }

      ${
        styleProps.alignY &&
        css`
          justify-content: ${FLEX_VERTICAL_ALIGN_MAP[styleProps.alignY]};
        `
      }

      ${
        styleProps.alignX &&
        css`
          align-items: ${FLEX_HORIZONTAL_ALIGN_MAP[styleProps.alignX]};
        `
      }
    `}
  `;
}
