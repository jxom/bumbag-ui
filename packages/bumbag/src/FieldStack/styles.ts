import { css, cssClass } from '../styled';
import { breakpoint, theme } from '../utils';

export const FieldStack = (styleProps) => cssClass`
  & {
    ${theme(styleProps.themeKey, `styles.base`)(styleProps)};
  }
`;
