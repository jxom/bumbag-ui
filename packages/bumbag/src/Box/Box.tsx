import * as React from 'react';
import { Box as ReakitBox, BoxProps as ReakitBoxProps } from 'reakit';

import { ThemeConfig, CSSProperties } from '../types';
import {
  useStyle,
  useClassName,
  omitCSSProps,
  pickHTMLProps,
  mergeRefs,
  createComponent,
  createElement,
  createHook,
} from '../utils';

import * as styles from './styles';

type ComponentType<R> = React.ComponentType<R> & { useProps: any };

export type LocalBoxProps = {
  use?: string | ComponentType<any>;
  className?: string;
  children?: React.ReactNode | ((props: BoxProps) => React.ReactNode);
  alignX?: 'left' | 'center' | 'right';
  alignY?: 'top' | 'center' | 'bottom';
  altitude?: string;
  variant?: string;
  colorMode?: string;
  /* Component-level theme overrides [Read more](/theming/#component-theming) */
  overrides?: ThemeConfig;
  elementRef?: React.Ref<any>;
  themeKey?: string;
};
export type BoxProps = ReakitBoxProps & CSSProperties & LocalBoxProps;

const useProps = createHook<BoxProps>(
  (_props, { themeKey, themeKeyOverride }) => {
    let props = _props;
    const { use } = props;

    if (use && typeof use !== 'string' && use.useProps) {
      const newProps = use.useProps({ ...props, use: undefined });
      props = { ...props, ...newProps };
    }

    // Convert CSS props to an object.
    // Example input:
    // props = { color: 'red', backgroundColor: 'blue', isEnabled: true }
    //
    // Example output:
    // style = { color: 'red', backgroundColor: 'blue' }
    const style = useStyle(props);

    // Append the styles from above as a className on the DOM element (with precedence).
    let className = useClassName({
      style: styles.style,
      styleProps: { ...props, style },
      themeKey,
      themeKeyOverride,
      prevClassName: props.className,
    });

    // Append the Box styles as a className on the DOM element.
    className = useClassName({
      style: styles.Box,
      styleProps: props,
      prevClassName: className,
      themeKey,
      themeKeyOverride,
    });

    // Pick out and invalid HTML props & omit the CSS props.
    const htmlProps = omitCSSProps(pickHTMLProps({ ...props, className }));

    if (props.elementRef) {
      // @ts-ignore
      htmlProps.ref = mergeRefs(props.elementRef, props.ref);
    }

    if (props.wrapElement) {
      // @ts-ignore
      htmlProps.wrapElement = props.wrapElement;
    }
    return { ...htmlProps };
  },
  { themeKey: 'Box' }
);

export const Box = createComponent<BoxProps>(
  (props) => {
    const boxProps = useProps(props);
    return createElement({ children: props.children, component: ReakitBox, use: props.use, htmlProps: boxProps });
  },
  {
    attach: { displayName: 'Box', useProps },
    themeKey: 'Box',
  }
);
