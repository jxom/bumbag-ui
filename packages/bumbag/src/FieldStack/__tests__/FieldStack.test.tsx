import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Box } from '../../Box';
import { FieldStack } from '../FieldStack';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(
      <FieldStack>
        <Box />
        <Box />
      </FieldStack>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(
      <FieldStack color="primary">
        <Box />
        <Box />
      </FieldStack>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with spacing prop', () => {
    const { container } = render(
      <FieldStack spacing="major-4">
        <Box />
        <Box />
      </FieldStack>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with orientation prop', () => {
    const { container } = render(
      <FieldStack orientation="horizontal">
        <Box />
        <Box />
      </FieldStack>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(
        <FieldStack use="div">
          <Box />
          <Box />
        </FieldStack>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with FieldStack props', () => {
      const { result } = renderHook(() => FieldStack.useProps());
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(
        <FieldStack>
          {(FieldStackProps) => (
            <div {...FieldStackProps}>
              <Box />
              <Box />
            </div>
          )}
        </FieldStack>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('theming', () => {
  it('FieldStack.base should render correctly', () => {
    const { container } = render(
      <FieldStack>
        <Box />
        <Box />
      </FieldStack>,
      {
        // @ts-ignore
        theme: { FieldStack: { styles: { base: { backgroundColor: 'red' } } } },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(
      <FieldStack>
        <Box />
        <Box />
      </FieldStack>,
      {
        // @ts-ignore
        theme: { FieldStack: { defaultProps: { className: 'test', color: 'primary' } } },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
