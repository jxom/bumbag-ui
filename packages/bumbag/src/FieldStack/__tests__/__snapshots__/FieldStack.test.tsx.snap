// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5.emotion-5 > *:not(:last-child) {
  margin-bottom: 1rem;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Stack fp-FieldStack"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`composition hook should return with FieldStack props 1`] = `
Object {
  "className": "css-fvskbx css-1vaxa0h css-4lfl7m css-1yuzjtp fp-Box fp-Stack fp-FieldStack",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5.emotion-5 > *:not(:last-child) {
  margin-bottom: 1rem;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Stack fp-FieldStack"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5.emotion-5 > *:not(:last-child) {
  margin-bottom: 1rem;
}

.emotion-7 {
  color: #574feb !important;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-7 test fp-Box fp-Stack fp-FieldStack"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`props should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5.emotion-5 > *:not(:last-child) {
  margin-bottom: 1rem;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Stack fp-FieldStack"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5.emotion-5 > *:not(:last-child) {
  margin-bottom: 1rem;
}

.emotion-7 {
  color: #574feb !important;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-7 fp-Box fp-Stack fp-FieldStack"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`props should render correctly with orientation prop 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

@media screen and (min-width:768px) {
  .emotion-5 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }

  .emotion-5 > * {
    -webkit-flex: 1;
    -ms-flex: 1;
    flex: 1;
  }

  .emotion-5 > *:not(:last-child) {
    margin-right: 1rem;
  }
}

@media screen and (max-width:768px) {
  .emotion-5.emotion-5 > *:not(:last-child) {
    margin-bottom: 1rem;
  }
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Stack fp-FieldStack"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`props should render correctly with spacing prop 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5.emotion-5 > *:not(:last-child) {
  margin-bottom: 2rem;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Stack fp-FieldStack"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`theming FieldStack.base should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5.emotion-5 > *:not(:last-child) {
  margin-bottom: 1rem;
}

.emotion-4.emotion-4 {
  background-color: #ff0000;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Stack fp-FieldStack"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;
